### PR TITLE
Merged deals

### PR DESCRIPTION
**Pull request by fivetran-reneeli**
## PR Overview
**This PR will address the following Issue/Feature:** https://github.com/fivetran/dbt_hubspot/issues/134

**This PR will result in the following new package version:** 0.14.1

**Please detail what change(s) this PR introduces and any additional information that should be known during the review of this PR:** 

We are adding the `merged_deals` table into our package to be able to add merge logic downstream to stale deals.

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt compile
- [ ] dbt run –full-refresh
- [x] dbt run
- [x] dbt test
- [x] dbt run –vars (if applicable) -- see height ticket for variables used

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked and tagged
- [x] You are assigned to the corresponding issue and this PR
- [x] BuildKite integration tests are passing

### Detailed Validation
Please acknowledge that the following validation checks have been performed prior to marking this PR as "ready for review":
- [x] You have validated these changes and assure this PR will address the respective Issue/Feature.
- [x] You are reasonably confident these changes will not impact any other components of this package or any dependent packages.
- [x] You have provided details below around the validation steps performed to gain confidence in these changes.

 see linked hex journal

### Standard Updates
Please acknowledge that your PR contains the following standard updates:
- Package versioning has been appropriately indexed in the following locations:
    - [x]  indexed within dbt_project.yml
    - [x] indexed within integration_tests/dbt_project.yml
- [x] CHANGELOG has individual entries for each respective change in this PR
    
- [x] README updates have been applied (if applicable)
    
- [ ] DECISIONLOG updates have been updated (if applicable)
- [x] Appropriate yml documentation has been added (if applicable)

### dbt Docs
Please acknowledge that after the above were all completed the below were applied to your branch:
- [ ] docs were regenerated (unless this PR does not include any code or yml updates)

### If you had to summarize this PR in an emoji, which would it be?
 
:dancer:

### CODE REVIEW

**Comment by fivetran-reneeli** [comment]
- regen docs

File: models/stg_hubspot__merged_deal.sql [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{{ config(enabled=fivetran_utils.enabled_vars(['hubspot_sales_enabled','hubspot_deal_enabled'])) }}
```
Comment (fivetran-joemarkiewicz) [inline models/stg_hubspot__merged_deal.sql:1]: Should we also consider adding a variable to have this disabled via a new variable if the `merged_deal` table is not present in the source?

Do we know how common this table is?

File: models/tmp/stg_hubspot__merged_deal_tmp.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{{ config(enabled=fivetran_utils.enabled_vars(['hubspot_sales_enabled','hubspot_deal_enabled'])) }}
```
Comment (fivetran-joemarkiewicz) [inline models/tmp/stg_hubspot__merged_deal_tmp.sql:1]: Same comment about possibly adding a new variable here for the new table. If it is low enough then we should create a new variable for this model. I could see it being common that not enough accounts leverage merged deals in HubSpot and we should create a new variable.

File: models/src_hubspot.yml [hunk 3]
```diff
@@ -1311,3 +1311,16 @@ sources:
             description: Short summary of ticket.
           - name: property_content
             description: Text in body of the ticket.
+
+      - name: merged_deal
+        identifier: '{{ var("hubspot_merged_deal_identifier", "merged_deal") }}'
+        description: Each record represents a deal that has been merged along with the related deal it's been merged into.
+        config:
+          enabled: "{{ var('hubspot_sales_enabled', true) and var('hubspot_deal_enabled', true) }}"
```
Comment (fivetran-joemarkiewicz) [inline models/src_hubspot.yml:1319]: Same comment that we will need to update this if we decide to create a new variable.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,10 @@
+# dbt_hubspot_source v0.14.1
+[PR #125](https://github.com/fivetran/dbt_hubspot_source/pull/125) includes the following updates:
+
+## Features
+- Adds the `stg_hubspot__merged_deal` model. Downstream, this model is used to remove stale deals and aggregate them to the deal they have been merged into. 
+  - This can be disabled by setting `hubspot_sales_enabled` and `hubspot_deal_enabled` to False in your dbt_project.yml. See [Step 4 of the README](https://github.com/fivetran/dbt_hubspot_source#step-4-disable-models-for-non-existent-sources) for more details.
```
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:6]: If we add a new variable we will want to document it here as well as within the README.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,10 @@
+# dbt_hubspot_source v0.14.1
```
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:1]: I am almost considering that this release should possibly be a breaking `v0.15.0` release.

The addition of a new model could be breaking for users that do not have this in their source. I imagine we should either:
- Keep this as non breaking, but add a new variable for the `merged_deal` and have it be `false` by default so it doesn't break for users.
- Make this a breaking change and also add the new variable, but have the variable be `true` by default.
- Do not create a variable and just make this a breaking change. However, we still run the risk of users upgrading and running into an issue if they do not have the source table.

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-reneeli thanks for working through this update! Generally your updates look great, I just have some suggestions and questions around possibly adding a variable for the `merged_deal` models and possibly making this a breaking change. 

Let me know if you have any questions. Thanks!

File: CHANGELOG.md (hunk 5 above)
Comment (fivetran-reneeli) [reply CHANGELOG.md:1]: This is a good point that not all users will have `merged_deal`. When thinking through the options, I'm keeping quickstart in mind and therefore I know it's best to be conservative with the addition of new variables. So I think the option that will prevent the most support tickets down the line would be having a variable for the new model which the default is false.

File: CHANGELOG.md (hunk 4 above)
Comment (fivetran-reneeli) [reply CHANGELOG.md:6]: done

File: models/src_hubspot.yml (hunk 3 above)
Comment (fivetran-reneeli) [reply models/src_hubspot.yml]: updated!

File: models/tmp/stg_hubspot__merged_deal_tmp.sql (hunk 2 above)
Comment (fivetran-reneeli) [reply models/tmp/stg_hubspot__merged_deal_tmp.sql]: added here as well

**Review by fivetran-reneeli — COMMENTED** [review]
Thanks Joe for calling out the fact that not all users will have merged_deal and therefore we should include a variable-- I think the most 'fail-proof' way would be having it as default false (I'm thinking about quickstart implications)

As such I'll keep the version as is since it will be non breaking

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-reneeli thanks for addressing my review notes. There are still a few gaps in the latest commits. Please take a look and address the remaining notes. Let me know if you have any questions. Thanks!

File: models/src_hubspot.yml [hunk 6]
```diff
@@ -1311,3 +1311,16 @@ sources:
             description: Short summary of ticket.
           - name: property_content
             description: Text in body of the ticket.
+
+      - name: merged_deal
+        identifier: '{{ var("hubspot_merged_deal_identifier", "merged_deal") }}'
+        description: Each record represents a deal that has been merged along with the related deal it's been merged into.
+        config:
+          enabled: "{{ var('hubspot_sales_enabled', true) and var('hubspot_deal_enabled', true) and var('hubspot_merged_deal_enabled', true) }}"
```
[the remark this answers is not in the document]
Comment (fivetran-reneeli) [reply models/src_hubspot.yml]: Thanks for catching-- updated the defaults to be false as expected

File: integration_tests/dbt_project.yml [hunk 7]
```diff
@@ -1,18 +1,19 @@
...
     hubspot_service_enabled: true
-    # hubspot_sales_enabled: true # enable when generating docs
-    # hubspot_marketing_enabled: true # enable when generating docs
-    # hubspot_contact_merge_audit_enabled: true # enable when generating docs
-    # hubspot_using_all_email_events: true # enable when generating docs
+    hubspot_sales_enabled: true # enable when generating docs
+    hubspot_marketing_enabled: true # enable when generating docs
+    hubspot_contact_merge_audit_enabled: true # enable when generating docs
+    hubspot_using_all_email_events: true # enable when generating docs
+    hubspot_merged_deal_enabled: true # enable when generating docs
```
[the remark this answers is not in the document]
Comment (fivetran-reneeli) [reply integration_tests/dbt_project.yml]: ah yes forgot to after generating docs, commented back out

File: README.md [hunk 8]
```diff
@@ -105,6 +105,7 @@ vars:
   hubspot_engagement_task_enabled: false
   hubspot_owner_enabled: false
   hubspot_property_enabled: false                         # Disables property and property_option tables
+  hubspot_merged_deal_enabled: true                       # Enables the merged_deal table, which will filter out merged deals from the final deal models.
```
[the remark this answers is not in the document]
Comment (fivetran-reneeli) [reply README.md]: moved to the other `deal` variables

File: CHANGELOG.md (hunk 5 above)
Comment (fivetran-reneeli) [reply CHANGELOG.md:1]: updated to default false, thanks for catching

File: README.md [hunk 9]
```diff
@@ -91,6 +91,7 @@ vars:
   hubspot_company_enabled: false
   hubspot_company_property_history_enabled: false         # Disable the company property history models
   hubspot_deal_enabled: false
+  hubspot_merged_deal_enabled: true                       # Enables the merged_deal table, which will be used to filter out merged deals from the final deal models. Note that `hubspot_sales_enabled` and `hubspot_deal_enabled` must not be set to False.
```
Comment (fivetran-joemarkiewicz) [inline README.md:94]:
```suggestion
  hubspot_merged_deal_enabled: true                       # Enables the merged_deal table, which will be used to filter out merged deals from the final deal models. False by default. Note that `hubspot_sales_enabled` and `hubspot_deal_enabled` must not be set to False.
```

File: models/stg_hubspot__deal.yml [hunk 10]
```diff
@@ -161,3 +161,18 @@ models:
...
+  - name: stg_hubspot__merged_deal
+    description: Each record represents a deal that has been merged along with the related deal it's been merged into.
+    columns:
+      - name: merged_deal_id
+        description: ID of deal that has been merged.
+        tests:
+          - unique
+          - not_null
+      - name: deal_id
+        description: The ID of the deal the record has been merged INTO.
```
Comment (fivetran-joemarkiewicz) [inline models/stg_hubspot__deal.yml:174]: Is this true? I feel there could be scenarios where this is not in fact unique. I think the following would be a more representative test for this model.
```suggestion
      - name: merged_deal_id
        description: ID of deal that has been merged.
        tests:
          - not_null
      - name: deal_id
        description: The ID of the deal the record has been merged INTO.
        tests:
          - not_null
```

**Review by fivetran-joemarkiewicz — APPROVED** [review]
@fivetran-reneeli thanks for applying the updates! I have some final suggestions. Let me know if you have any questions. If not, then this is good to go!

File: models/stg_hubspot__deal.yml (hunk 10 above)
Comment (fivetran-reneeli) [reply models/stg_hubspot__deal.yml:174]: @fivetran-joemarkiewicz non-unique scenarios as in where a merged deal could have been merged into more than 1 deal and therefore show up more than once in the merged_deal table? Updating the tests make sense but just wondering if you had any specifics that popped into mind

**Review by fivetran-avinash — APPROVED** [review]

**Pull request merged**